### ISSUE TITLE
Definition of transposition

### ISSUE THREAD

**Issue by snowleopard** [conal/linalg#35]
I've seen transposition mentioned here a few times. Do we have a definition of what it means for linear maps?

While looking at #34, I remembered one curious definition that I learned from @simonpj.

Let `m : X ⊸ Y` be a linear map, `m @ x` stand for application of `m` to a vector `x`, and `x • y` stand for dot product of vectors `x` and `y`. Then, `transpose : (X ⊸ Y) -> (Y ⊸ X)` can be _uniquely_ defined by the following property:

```
∀ x : X, y : Y. (transpose m @ y) • x = y • (m @ x)
```

While this definition might seem somewhat mysterious, I like it because it doesn't involve any dual vector spaces, and can be used to mechanically prove that the natural implementation of `transpose` is correct:

```haskell
transpose :: L c r s -> L r c s
transpose (Scale s)  = Scale s
transpose (f :|# g)  = tr f :&# tr g
transpose (f :&# g)  = tr f :|# tr g
transpose (ForkL ms) = JoinL (fmap transpose ms)
transpose (JoinL ms) = ForkL (fmap transpose ms)
```

For example, here is a short proof that `transpose (Scale s) = Scale s` satisfies the above definition:

```
(transpose (Scale s) @ y) • x = (Scale s @ y) • x = y • (Scale s @ x)
```

The second step is a property of dot product and scalar multiplication.

By the way, it seems that there is currently no definition of `@`, i.e. the meaning of a linear map, in the repository. One might say, that composition `.@` does in fact define this meaning (since application and composition of linear maps are pretty much the same thing) but it may still be useful to have `@` written down explicitly, for example, to make it easier to prove that all other equations in the above implementation of `transpose` match the definition.

**Comment by bolt12** [comment]
Oh that's interesting! Do you have any reference to that property and to why does it _uniquely_ defines `transpose`? 

I also think that a lot of discussions around `transposition` and other operations revolve more around their meaning in terms of the linear map denotation and so, denotationally, the meaning of linear map transposition is the one expressed in my code (#34 ) (I think) and from that we just need/can go back to its matrix representation which is what I'm trying to do in `trT2` at #34. So `trT` has the correct type signature wrt to the denotation and denotes the essence of the meaning of linear map transposition and `trT2` tries to go from that to a more familiar type signature. I hope what I'm saying makes sense, I'm still new to the whole "denotational design" headspace. 

I'm also trying to implement `@` and am running into a wall:

```haskell
(@@) :: (Semiring s) => L f g s -> f s -> g s
Scale s @@ gs   = liftR2 (*) (pureRep s) gs
(p :&# q) @@ gs = p @@ gs :*: q @@ gs
(p :|# q) @@ gs = _
ForkL ms @@ gs  = Comp1 (fmap (@@ gs) ms)
JoinL ms @@ gs  = _
```

I don't know what to do in those two cases... it feels like I need to partition `gs` somehow, any help?

**Comment by dwincort** [comment]
First off, the word `transpose` is very strongly linked with matrices, and it may not be the best name if we want to be thinking about linear maps.  How about we start calling the function we're trying to define `dual`?

Second, I think the key to getting `(@@)` to work is to recognize that when `L f g s` matches to `(p :|# q)`, then we know that the type variable `f` is equal to `f' :*: h` for some `f'` and `h`.  We can then use this when we pattern match on the second argument.  For instance:
```haskell
(@@) :: Semiring s => L f g s -> f s -> g s
Scale s   @@ x            = (* s) <$> x
(p :&# q) @@ fs           = p @@ fs :*: q @@ fs
(p :|# q) @@ (fs :*: fs') = liftR2 (+) (p @@ fs) (q @@ fs')
ForkL ms  @@ fs           = Comp1 (fmap (@@ fs) ms)
JoinL ms  @@ (Comp1 fs)   = foldr (liftR2 (+)) (pureRep zero) $ liftR2 (@@) ms fs
```
I feel like there may be room for more simplification (perhaps using an `Additive` constraint in the right way), but I don't have time to play with it right now.

**Comment by bolt12** [comment]
Yes I agree! I'll change the names accordingly. And thanks a lot for the help on `(@@)`

**Comment by snowleopard** [comment]
> Oh that's interesting! Do you have any reference to that property and to why does it uniquely defines transpose?

@bolt12 Simon gave a talk "Automatic Differentiation for Dummies" which used the above definition of `transpose`. The recorded version that I found online (https://www.youtube.com/watch?v=FtnkqIsfNQc, around 27:50) uses a slightly different variant based on "reverse application" with a similar property (the reverse application operation is also uniquely defined by the property).

Why are `tranpose` and reverse application operations defined uniquely by these properties? It's not immediately obvious, but it's not difficult to prove that if there are two functions `transpose1` and `transpose2` that satisfy the property, then `transpose1 @ x == transpose2 @ x` for all possible `x : X`, i.e. they are extensionally equal (the proof is not too difficult, I can recall it if need be).

> First off, the word transpose is very strongly linked with matrices, and it may not be the best name if we want to be thinking about linear maps. How about we start calling the function we're trying to define dual?

@dwincort Indeed, transpose comes from the world of matrices. One could use `dual`, or "reverse application" as in Simon's talk.

Still, the underlying question remains: Is there a definition of what we mean by `transpose`, `dual`, or `reverseAp` in the general context of linear maps? Does the above definition seem sensible? Are there better alternatives?

**Comment by capn-freako** [comment]
A very fuzzy and not very complete thought:

In Haskell, `sequence` is synonymous w/ `transpose` (at least, for nested functors, which seems an apropos context to us).
And the thought of reversing a sequence seems to have a lot of overlap with that of "reverse application" of linear maps (transforms).
